core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -15,7 +15,67 @@ import scala.concurrent.duration._
 
 /**
   * A stream producing output of type `O` and which may evaluate `F`
-  * effects. If `F` is [[Pure]], the stream evaluates no effects.
+  * effects.
+  * Or: a lazy sequence of computations that may evaluate `F` effects `F` and yield values of type `O`.
+  *
+  * Streams are purely functional: a `Stream[F, O]` is a description for a computation,
+  * much like `cats.effect.IO` or `scalaz.Task` are just descriptions of computations.
+  * The methods in the `Stream` class only calculate new descriptions from others.
+  *
+  * Streams are lazy, pull-based data sources. To '''evaluate''' a stream is to extract from it
+  * as many elements as a _consumer_ of the stream may need for the result.
+  *   by performing as many `F`-effectful computations as necessary to get those elements.
+  *   Thus, the effects in `F` are not evaluated unless and until they are required (pulled) by a stream consumer.
+  *
+  * - '''Laziness''': streams are lazily evaluated sources for effectful computations.
+  *   The operations on a Stream only evaluate a finite prefix,
+  *   a finite part of a whole stream, as needed to supply the requests of the stream _consumers_.
+  *   Thus, there can be infinite streams, or Streams based on
+  *   an unlimited source of information (e.g. a Signal that may be sampled with as frequently as desired).
+  *
+  * - '''Source''': a stream may simply be a fixed list of pre-defined F-effectful computations.
+  *   However, it can also be backed by an external data source (e.g. a network connection).
+  *   It can also be backed by a effectful computation that can be re-evaluated,
+  *   such as a Stream to read the current system time.
+  *
+  * There are some special properties or cases of streams:
+  *  - A '''pure''' stream is a `Stream` in which the `F` is [[Pure]], which evaluates no effects.
+  *  - A stream is '''terminating''', or '''finite''', if pulling from it reaches the end after yielding
+  *    a finite number of elements.
+  *  - In particular, a stream is ''''empty''' if it terminates after yielding no value.
+  *  - A '''singleton''' stream is a stream that ends after yielding one single value.
+  *  - A '''never''' stream is a stream that never terminates and never yields any value.
```

Comment:
Is `s.drain` a never stream using this classification system? I.e. s may evaluate arbitrary effects forever, never emitting a value.

Reply:
Yes.

I used _never_ after the `Stream.never` method, but I see that evaluation of effects is orthogonal to yielding values, so there can be a stream with unlimited valueless effects, or with effects with values. IIUC, however, a stream can never yield a value without at least performing one effect, so every yield value comes out of an effect.

The reason I want to differentiate between a _never_-yielding computation and an infinite one is that  lazy operations, like `take`, always yield values when applied to infinite non-failing stream; others, like `exists` may not yield a value out of an infinite stream or not; and others, like `fold`, yield nothing until they reach the stream's end.